File: thcon/src/app/gnome_shell.rs
//! Switches between [GNOME Shell](https://wiki.gnome.org/Projects/GnomeShell) user themes, like
//! the [User Themes extension](https://extensions.gnome.org/extension/19/user-themes/) does
//!
//! ## Usage: Linux & BSD
//! GNOME Shell user themes require the [User Themes
//! extension](https://extensions.gnome.org/extension/19/user-themes/) to be enabled.  Once that's
//! done, simply provide the name of the theme as displayed in the User Themes extension config
//! (either via GNOME Extensions or GNOME Tweaks), e.g.:
//!
//! ```toml
//! [gnome-shell]
//! light = "Arc"
//! dark = "Arc-Dark-solid"
//! ```
//!
//! ## Usage: Windows & macOS
//! Currently unsupported.
//!
//! ## `thcon.toml` Schema
//! Section: `gnome-shell`
//!
//! | Key | Type | Description | Default |
//! | --- | ---- | ----------- | ------- |
//! | `disabled` | boolean | `true` to disable theming of this app, otherwise `false` | `false` |
//! | `dark` | string | The name of the theme (case-sensitive) to apply in dark mode | (none) |
//! | `light` | string | The name of the theme (case-sensitive) to apply in light mode | (none) |
//!

use crate::config::Config as ThconConfig;
use crate::operation::Operation;
use crate::themeable::{ConfigError, ConfigState, Themeable};
use crate::AppConfig;
use crate::Disableable;

use anyhow::{Context, Result};
use gio::SettingsExt;
use serde::Deserialize;

#[derive(Debug, Deserialize, Disableable, AppConfig)]
pub struct _Config {
    light: String,
    dark: String,
    #[serde(default)]
    disabled: bool,
}

pub struct GnomeShell {}

impl Themeable for GnomeShell {
    fn config_state(&self, config: &ThconConfig) -> ConfigState {
        ConfigState::with_manual_config(config.gnome_shell.as_ref().map(|c| c.inner.as_ref()))
    }

    fn switch(&self, config: &ThconConfig, operation: &Operation) -> Result<()> {
        let config = match self.config_state(config) {
            ConfigState::NoDefault => {
                return Err(ConfigError::RequiresManualConfig("gnome_shell").into())
            }
            ConfigState::Default => unreachable!(),
            ConfigState::Disabled => return Ok(()),
            ConfigState::Enabled => config.gnome_shell.as_ref().unwrap().unwrap_inner_left(),
        };

        let theme = match operation {
            Operation::Darken => &config.dark,
            Operation::Lighten => &config.light,
        };

        let gsettings = gio::Settings::new("org.gnome.shell.extensions.user-theme");
        gsettings
            .set_string("name", theme)
            .map(|_| gio::Settings::sync())
            .with_context(|| format!("Unable to apply GNOME Shell user theme '{}'", theme))
    }
}

File: lib/apps/gnome_shell.go
//go:build linux || dragonfly || freebsd || netbsd || openbsd

package apps

import (
	"context"

	"github.com/gotk3/gotk3/glib"
	"github.com/theme-controller/thcon/lib/health"
	"github.com/theme-controller/thcon/lib/operation"
)

type GnomeShellConfig struct {
	GnomeShell *struct {
		Disabled bool   `toml:"disabled"`
		Dark     string `toml:"dark" validate:"required_with=Light"`
		Light    string `toml:"light" validate:"required_with=Dark"`
	} `toml:"gnome-shell"`
}

type GnomeShell struct{}

var _ Switchable = (*GnomeShell)(nil)

func (g *GnomeShell) ValidateConfig(ctx context.Context, config *Config) (health.Status, error) {
	return health.RequiresConfig(ctx, config.GnomeShell)
}

func (g *GnomeShell) Switch(ctx context.Context, mode operation.Operation, config *Config) error {
	gsettings := glib.SettingsNew("org.gnome.shell.extensions.user-theme")
	var theme = "Arc-Dark-solid"
	if mode == operation.LightMode {
		theme = "Arc"
	}
	gsettings.SetString("name", theme)
	glib.SettingsSync()

	return nil
}

func (gt *GnomeShell) Name() string {
	const name = "GNOME Shell User Theme"
	return name
}

func (gt *GnomeShell) Argname() string {
	const argname = "gnome-shell"
	return argname
}
